packages/util/src/emulator.ts

```diff
@@ -140,3 +140,103 @@ export function createMockUserToken(
     signature
   ].join('.');
 }
+
+interface EmulatorStatuses {
+  [name: string]: boolean;
+}
+const emulatorStatus: EmulatorStatuses = {};
+
+export interface EmulatorStatus {
+  isRunningEmulator: boolean;
+}
+
+// Checks whether any products are running on an emulator
+function areRunningEmulator(): boolean {
+  let runningEmulator = false;
+  for (const key of Object.keys(emulatorStatus)) {
+    if (emulatorStatus[key]) {
+      runningEmulator = true;
+    }
+  }
+  return runningEmulator;
+}
+
+function getOrCreateEl(id: string): { created: boolean; element: HTMLElement } {
+  let parentDiv = document.getElementById(id);
+  let created = false;
+  if (!parentDiv) {
+    parentDiv = document.createElement('div');
+    parentDiv.setAttribute('id', id);
+    created = true;
+  }
+  return { created, element: parentDiv };
+}
+
+/**
+ * Updates Emulator Banner. Primarily used for Firebase Studio
+ * @param name
+ * @param isRunningEmulator
+ * @public
+ */
+export function updateEmulatorBanner(
+  name: string,
+  isRunningEmulator: boolean
+): void {
+  if (emulatorStatus[name] === isRunningEmulator) {
+    // No rerendering required
+    return;
+  }
+  emulatorStatus[name] = isRunningEmulator;
+  const bannerId = '__firebase__banner';
+  if (!areRunningEmulator()) {
+    tearDown();
+    return;
+  }
+
+  function tearDown(): void {
+    if (typeof document !== 'undefined') {
+      const element = document.getElementById(bannerId);
+      if (element) {
+        element.remove();
+      }
+    }
+  }
+
+  function setupDom(): void {
+    const banner = getOrCreateEl(bannerId);
+    const firebaseText: HTMLSpanElement =
+      document.getElementById('__firebase__text') ||
+      document.createElement('span');
+    if (banner.created) {
+      // update styles
+      const bannerEl = banner.element;
+      bannerEl.style.display = 'flex';
+      bannerEl.style.background = '#7faaf0';
+      bannerEl.style.position = 'absolute';
+      bannerEl.style.bottom = '5px';
+      bannerEl.style.left = '5px';
+      bannerEl.style.padding = '.5em';
+      bannerEl.style.borderRadius = '5px';
+      bannerEl.style.alignContent = 'center';
+      const closeBtn = document.createElement('span');
+      closeBtn.style.cursor = 'pointer';
+      closeBtn.style.paddingLeft = '5px';
+      closeBtn.innerHTML = ' &times;';
+      closeBtn.onclick = () => {
+        tearDown();
+      };
+      bannerEl.appendChild(firebaseText);
+      bannerEl.appendChild(closeBtn);
+      document.body.appendChild(banner.element);
+    }
+    firebaseText.setAttribute('id', '__firebase__text');
+    firebaseText.innerText = 'Running in this workspace';
+  }
+  if (typeof window !== 'undefined' && typeof document !== 'undefined') {
```

Comment:
Can we put this check at the top of `updateEmulatorBanner` and have it return early if these are undefined? Then we won't have to check for document in  `tearDown()` and we can't accidentally add code in the future we forget to check.

Reply:
Done